compiler/src/dmd/expressionsem.d

```diff
@@ -14771,11 +14771,14 @@ private bool checkArithmetic(Expression e, EXP op)
         return true;
     }
 
-    // FIXME: Existing code relies on adding / subtracting types in typeof() expressions:
-    // alias I = ulong; alias U = typeof(I + 1u);
-    // https://github.com/dlang/dmd/issues/20763
-    if (op == EXP.add || op == EXP.min)
+    if ((op == EXP.add || op == EXP.min) && e.isTypeExp())
+    {
+        // @@@DEPRECATED_2.121@@@
+        // Deprecated in 2.111
+        // In 2.121, remove this branch to let `checkValue` raise the error
+        deprecation(e.loc, "type `%s` has no value", e.toChars);
```

Comment:
can we add a supplemental to do (possibly after checking we are in `typeof`) that says 
```
"use `%s.init %s %s` instead", e.toChars, ExpToString(op), rhs"
```

Reply:
I added one, but it's not as detailed as you suggested because then I have to add BinExp and Scope parameters to `checkValue` which makes it more complex.

Reply:
and it would be good if we could add that to the error case for the other operators too.

Reply:
> I added one, but it's not as detailed as you suggested because then I have to add BinExp and Scope parameters to checkValue which makes it more complex.

That's not going to give misleading results w.r.t integral promotions? I don't think they should occur here. But that is probably good enough.

Reply:
I don't understand what you mean?

Reply:
In which case I should wake up more before asking silly questions.